packages/core/cyclist.mjs

```diff
@@ -1,93 +1,136 @@
 /*
-cyclist.mjs - <short description TODO>
+cyclist.mjs - recieves clock pulses from clockworker, and schedules the next events
 Copyright (C) 2022 Strudel contributors - see <https://github.com/tidalcycles/strudel/blob/main/packages/core/cyclist.mjs>
 This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details. You should have received a copy of the GNU Affero General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
 
-import createClock from './zyklus.mjs';
 import { logger } from './logger.mjs';
 
 export class Cyclist {
-  constructor({ interval, onTrigger, onToggle, onError, getTime, latency = 0.1 }) {
+  constructor({ onTrigger, onToggle, getTime }) {
     this.started = false;
     this.cps = 0.5;
-    this.num_ticks_since_cps_change = 0;
     this.lastTick = 0; // absolute time when last tick (clock callback) happened
-    this.lastBegin = 0; // query begin of last tick
-    this.lastEnd = 0; // query end of last tick
     this.getTime = getTime; // get absolute time
-    this.num_cycles_since_last_cps_change = 0;
+
+    this.num_cycles_at_cps_change = 0;
     this.onToggle = onToggle;
-    this.latency = latency; // fixed trigger time offset
-    this.clock = createClock(
-      getTime,
-      // called slightly before each cycle
-      (phase, duration, tick) => {
-        if (tick === 0) {
-          this.origin = phase;
+    this.latency = 0.1; // fixed trigger time offset
+    this.cycle = 0;
+
+    this.worker = new SharedWorker(new URL('./clockworker.js', import.meta.url));
+    this.worker.port.start();
+
+    this.channel = new BroadcastChannel('strudeltick');
+    let worker_time_dif = 0; // time difference between audio context clock and worker clock
+    let weight = 0; // the amount of weight that is applied to the current average when averaging a new time dif
+    const maxWeight = 400;
+    const precision = 10 ** 3; //round off time diff to prevent accumulating outliers
+
+    // the clock of the worker and the audio context clock can drift apart over time
+    // aditionally, the message time of the worker pinging the callback to process haps can be inconsistent.
+    // we need to keep a rolling weighted average of the time difference between the worker clock and audio context clock
+    // in order to schedule events consistently.
+    const setTimeReference = (time, workertime) => {
```

Comment:
this looks like something that could potentially be solved with https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/getOutputTimestamp

Reply:
yes I think that could maybe help with the issue of audio context time being inconsistent if there are audio drops.

Reply:
After playing around with it, Im not sure how helpful have the performance time of the audio context, because we still need to schedule with a comparison between the worker clock relative to the audio ctx clock

Reply:
maybe there is another way to solve this problem

Reply:
I think they would be different because performance.now() on worker would be the active time since the worker was started, and the ctx performance time should be the time the ctx has been alive

Reply:
there is performance.timeOrigin to know when the origin of performance.now(), so the worker could potentially be given its creation time.. but we can also tackle that in a future PR